plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectRepository.java

```diff
@@ -103,7 +96,32 @@ private static Projects getProjectsApi(Credential credential) {
   }
 
   private static GcpProject convertToGcpProject(Project project) {
+    Preconditions.checkNotNull(project);
     return new GcpProject(project.getName(), project.getProjectId());
   }
 
+  /**
+   * @return true if the credential has access to the GCP project identified by {@code projectId}
+   * and the project has App Engine application
+   * @throws ProjectRepositoryException if an error other than HTTP 404 happens while retrieving the
+   * App Engine application
+   */
+  public boolean hasAppEngineApplication(Credential credential, String projectId)
```

Comment:
We call this lazily, so in theory we only do at most one call for the selected project (or none if there's no persisted selection) when opening the dialog. In practice we do 3 calls due to databinding triggering the selection changed event. This is one of the reasons for #1374 

Once the dialog is open, then if the user selects a project by clicking on it (or moving the selection with keys) we do one call.